<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>ORBIT for E. coli</title>

<script src="index_files/header-attrs-2.18/header-attrs.js"></script>
<script src="index_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="index_files/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="index_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="index_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="index_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="index_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="index_files/navigation-1.1/tabsets.js"></script>
<link href="index_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="index_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">ORBIT for <em>E. coli</em></h1>
<h3 class="subtitle">Kilobase-scale oligonucleotide recombineering at
high throughput and high efficiency</h3>

</div>


<p><br></p>
<center>
<p><em>This is a paper website for our manuscript. It is a convenient
wrapper for the github code repo, which contains html versions of all
code notebooks. Our goal is to help others use ORBIT, so we also link to
several resources that may be of interest. We welcome any feedback -
Thank you!</em></p>
<p><em>-Scott Saunders, Distinguished Fellow, UTSW Medical
Center</em></p>
<p><a
href="https://www.biorxiv.org/content/10.1101/2023.06.28.546561v1">bioRxiv
preprint</a>   |   <a href="github.com/saunders-lab/ecoli_orbit">github
repository</a>   | <a
href="https://saunderslab.org/research/ec_orbit/">Saunders lab ORBIT
help page</a>   |   <a
href="https://saunders-lab.shinyapps.io/ORBIT_TO_design_ecMG1655/">Targeting
oligo design app</a></p>
</center>
<div id="abstract" class="section level1">
<h1>Abstract</h1>
<hr />
<p>Microbiology and synthetic biology depend on reverse genetic
approaches to manipulate bacterial genomes; however, existing methods
require molecular biology to generate genomic homology, suffer from low
efficiency, and are not easily scaled to high throughput applications.
To overcome these limitations, we developed a system for creating
kilobase-scale genomic modifications that uses DNA oligonucleotides to
direct the integration of a non-replicating plasmid. This method,
<strong>O</strong>ligonucleotide <strong>R</strong>ecombineering
followed by <strong>B</strong>xb-1 <strong>I</strong>ntegrase
<strong>T</strong>argeting (<strong>ORBIT</strong>) was pioneered in
<em>Mycobacteria</em>, and here we adapt and expand it for <em>E.
coli</em>. Our redesigned plasmid toolkit achieved nearly 1000x higher
efficiency than λ Red recombination and enabled precise, stable
knockouts (&lt;134 kb) and integrations (&lt;11 kb) of various sizes.
Additionally, we constructed multi-mutants (double and triple) in a
single transformation, using orthogonal attachment sites. At high
throughput, we used pools of targeting oligonucleotides to knock out
nearly all known transcription factor and small RNA genes, yielding
accurate, genome-wide, single mutant libraries. By counting genomic
barcodes, we also show ORBIT libraries can scale to thousands of unique
members (&gt;30k). This work demonstrates that ORBIT for <em>E.
coli</em> is a flexible reverse genetic system that facilitates rapid
construction of complex strains and readily scales to create
sophisticated mutant libraries.</p>
<p><br></p>
</div>
<div id="code-notebooks" class="section level1">
<h1>Code notebooks</h1>
<hr />
<div id="main-figures" class="section level2">
<h2>Main figures</h2>
<ul>
<li><p>Figure 1. ORBIT overview and proof of principle. (No
computational content)</p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/main_figs/fig_2_protocol_optimization.html">Figure
2. ORBIT protocol optimization.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/main_figs/fig_3_gold_stds.html">Figure
3. Benchmarking ORBIT.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/main_figs/fig_4_sizes.html">Figure
4. Large deletions and insertions with ORBIT.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/main_figs/fig_5_multi_orbit.html">Figure
5. Orthogonal double mutants.</a></p></li>
<li><p>Figure 6. Markerless and scarless ORBIT strategies. (No
computational content)</p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/main_figs/fig_7_BC_oPool.html">Figure
7. Barcoded and multi-locus mutant libraries.</a></p></li>
<li><p><a href="TBD">Figure 8. High throughput knockout libraries for
transcription factor and small RNA genes.</a> link TBD</p></li>
</ul>
</div>
<div id="supplemental-figures" class="section level2">
<h2>Supplemental figures</h2>
<ul>
<li><p>Figure S1. Target oligo design principles and web app. (No
computational content)</p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/supplement_protocol_optimization.html">Figure
S2. Commercial oligo parameters and pricing.</a></p></li>
<li><p>Figure S3. Troubleshooting attB only colonies. (No computational
content)</p></li>
<li><p>Figure S4. Single gene deletion accuracy and stability. (No
computational content)</p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/supp_mutation_rate.html">Figure
S5. Helper plasmid efficiency and off target SNP mutation
rate.</a></p></li>
<li><p>Figure S6. Variable deletion size phenotypic accuracy. (No
computational content)</p></li>
<li><p>Figure S7. Sanger sequence confirmation of <em>galK</em>
deletions of various sizes. (No computational content)</p></li>
<li><p>Figure S8. Sanger sequence confirmation of <em>hisA, metA</em>
and <em>leuD</em> deletions of various sizes. (No computational
content)</p></li>
<li><p>Figure S9. Phenotypic accuracy of orthogonal and untargeted
double mutants. (No computational content)</p></li>
<li><p>[Figure S10. Orthogonal att sites and triple mutant
construction.]</p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/mut_lib_grids.html">Figure
S11. oPool library - Upstream and downstream reads for target
loci.</a></p></li>
<li><p>Figure S12. Low abundance (Twist Biosciences) targeting oligo
library processing. (No computational content)</p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/mut_lib_grids.html">Figure
S13. Short transcription factor library - Upstream and downstream reads
for target loci.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/mut_lib_grids.html">Figure
S14. Long transcription factor deletion library - Upstream and
downstream reads for target loci.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/mut_lib_grids.html">Figure
S15. Small RNA deletion library - Upstream and downstream reads for
target loci.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/supp_twist_lib.html">Figure
S16. Explaining mutant library abundances.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/supp_twist_lib.html">Figure
S17. Correlations of oligo parameters and observed reads per target
(perfect reads).</a></p></li>
</ul>
</div>
<div id="sequencing-data" class="section level2">
<h2>Sequencing data</h2>
<div id="galk-bc" class="section level3">
<h3>galK BC</h3>
<ul>
<li><a
href="https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/galK_BC_16N/galK_BC_16N_processing_analysis.html">galK
random genomic barcodes - sequencing data analysis.</a></li>
</ul>
</div>
<div id="twist-oligo-design" class="section level3">
<h3>Twist oligo design</h3>
<ul>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/twist_oligo_design/twist_orbit_TF_deletion.html">Twist
oligo pool design - Transcription factors</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/twist_oligo_design/twist_orbit_small_RNA_deletion.html">Twist
oligo pool design - Small RNAs</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/twist_oligo_design/twist_orbit_cloning_scheme.html">Twist
oligo pool design - Primers and cloning scheme</a></p></li>
</ul>
</div>
<div id="mutant-library-characterization" class="section level3">
<h3>Mutant library characterization</h3>
<ul>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/left_side_TO_libs_processing.html">Mutant
libraries - Junction #1 sequencing data processing.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/right_side_TO_libs_processing.html">Mutant
libraries - Junction #2 sequencing data processing.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/left_side_TO_libs_analysis.html">Mutant
libraries - Junction #1 analysis.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/right_side_TO_libs_analysis.html">Mutant
libraries - Junction #2 analysis.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/upstream_downstream_combined_analysis.html">Mutant
libraries - Combined junction 1 &amp; 2 analysis.</a></p></li>
</ul>
</div>
<div id="twist-to-abundance" class="section level3">
<h3>Twist TO abundance</h3>
<ul>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/TO_abundance/TO_abundance_processing.html">Twist
TO abundance - Sequencing data processing.</a></p></li>
<li><p><a
href="https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/TO_abundance/TO_abundance_analysis.html">Twist
TO abundance - Analysis.</a></p></li>
</ul>
<p><br></p>
</div>
</div>
</div>
<div id="other-resources" class="section level1">
<h1>Other resources</h1>
<hr />
<div id="protocols" class="section level2">
<h2>Protocols</h2>
<ul>
<li><a href="saunderslab.org/research/ec_orbit/protocols">Saunders lab
ORBIT protocols.</a></li>
</ul>
</div>
<div id="dna-maps" class="section level2">
<h2>DNA maps</h2>
<ul>
<li><p><a
href="https://github.com/saunders-lab/ecoli_orbit/tree/main/data/DNA_maps/plasmids">Plasmid
maps</a></p></li>
<li><p><a
href="https://github.com/saunders-lab/ecoli_orbit/tree/main/data/DNA_maps/genomic_loci">Genomic
loci</a></p></li>
<li><p><a
href="https://github.com/saunders-lab/ecoli_orbit/tree/main/data/DNA_maps/sequencing_libraries">Sequencing
libraries</a></p></li>
</ul>
</div>
<div id="addgene-plasmids" class="section level2">
<h2>Addgene plasmids</h2>
<ul>
<li><a href="TBD">Addgene page.</a> link TBD</li>
</ul>
</div>
<div id="sra-raw-sequencing-data" class="section level2">
<h2>SRA raw sequencing data</h2>
<ul>
<li><a
href="https://www.ncbi.nlm.nih.gov/sra/PRJNA989253">BioProject</a></li>
</ul>
<table>
<colgroup>
<col width="41%" />
<col width="19%" />
<col width="23%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th>Experiment</th>
<th align="center">Library Name</th>
<th align="center">Run ID and Link</th>
<th align="center">Size</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>galK barcodes</td>
<td align="center">galK_BC16N</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076308">SRR25076308</a></td>
<td align="center">123 Mb</td>
</tr>
<tr class="even">
<td>IDT oPool - left junction</td>
<td align="center">oPool_J1</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076307">SRR25076307</a></td>
<td align="center">8.3 Mb</td>
</tr>
<tr class="odd">
<td>Twist short TF deletions - left junction</td>
<td align="center">short_TF_J1</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076303">SRR25076303</a></td>
<td align="center">25 Mb</td>
</tr>
<tr class="even">
<td>Twist long TF deletions - left junction</td>
<td align="center">long_TF_J1</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076301">SRR25076301</a></td>
<td align="center">16 Mb</td>
</tr>
<tr class="odd">
<td>Twist small RNA deletions - left junction</td>
<td align="center">small_RNA_J1</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076299">SRR25076299</a></td>
<td align="center">18 Mb</td>
</tr>
<tr class="even">
<td>IDT oPool - right junction</td>
<td align="center">oPool_J2</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076304">SRR25076304</a></td>
<td align="center">15 Mb</td>
</tr>
<tr class="odd">
<td>Twist short TF deletions - right junction</td>
<td align="center">short_TF_J2</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076302">SRR25076302</a></td>
<td align="center">43 Mb</td>
</tr>
<tr class="even">
<td>Twist long TF deletions - right junction</td>
<td align="center">long_TF_J2</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076300">SRR25076300</a></td>
<td align="center">39 Mb</td>
</tr>
<tr class="odd">
<td>Twist small RNA deletions - right junction</td>
<td align="center">small_RNA_J2</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076298">SRR25076298</a></td>
<td align="center">30 Mb</td>
</tr>
<tr class="even">
<td>Twist short TF oligo abundance</td>
<td align="center">short_TF_oligos</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076297">SRR25076297</a></td>
<td align="center">45 Mb</td>
</tr>
<tr class="odd">
<td>Twist long TF oligo abundance</td>
<td align="center">long_TF_oligos</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076306">SRR25076306</a></td>
<td align="center">27 Mb</td>
</tr>
<tr class="even">
<td>Twist small RNA oligo abundance</td>
<td align="center">small_RNA_oligos</td>
<td align="center"><a
href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR25076305">SRR25076305</a></td>
<td align="center">42 Mb</td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
