---
title: "Figure 8: TO twist libraries"
subtitle: 'E. coli ORBIT 2022'
author: 'Scott H. Saunders'
output:
  html_document:
    theme: cosmo
    highlight: tango
    code_folding: show
    toc: yes
---

# Notes

This figure contains data from three ORBIT experiments where mutant libraries were created with mixtures of targeting oligos (from Twist Biosciences). Data figures were made in R notebooks and exported as pdfs. Cosmetic improvements were made in Adobe Illustrator. Note the diagram for Figures 8A was made in Adobe Illustrator.

See related code notebooks:

Supplemental figures

-   [Figure S13-S15. Twist mutant libraries - upstream and downstream reads for target loci.](https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/mut_lib_grids.html)

-   [Figure S16-S17. Explaining mutant library abundances.](https://saunders-lab.github.io/ecoli_orbit/code/supp_figs/supp_twist_lib.html)

Twist oligo design

-   [Twist oligo pool design - Transcription factors](https://saunders-lab.github.io/ecoli_orbit/code/twist_oligo_design/twist_orbit_TF_deletion.html)

-   [Twist oligo pool design - Small RNAs](https://saunders-lab.github.io/ecoli_orbit/code/twist_oligo_design/twist_orbit_small_RNA_deletion.html)

-   [Twist oligo pool design - Primers and cloning scheme](https://saunders-lab.github.io/ecoli_orbit/code/twist_oligo_design/twist_orbit_cloning_scheme.html)

Mutant library characterization

-   [Mutant libraries - Junction #1 sequencing data processing.](https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/left_side_TO_libs_processing.html)

-   [Mutant libraries - Junction #2 sequencing data processing.](https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/right_side_TO_libs_processing.html)

-   [Mutant libraries - Junction #1 analysis.](https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/left_side_TO_libs_analysis.html)

-   [Mutant libraries - Junction #2 analysis.](https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/right_side_TO_libs_analysis.html)

-   [Mutant libraries - Combined junction 1 & 2 analysis.](https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/upstream_downstream_combined_analysis.html)

Twist TO abundance

-   [Twist TO abundance - Sequencing data processing.](https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/TO_abundance/TO_abundance_processing.html)

-   [Twist TO abundance - Analysis.](https://saunders-lab.github.io/ecoli_orbit/code/seq_data_processing/TO_abundance/TO_abundance_analysis.html)

------------------------------------------------------------------------

Setup packages and plotting for the notebook:

```{r setup, echo=T, message=FALSE, warning=FALSE}
# Check packages
source("../tools/package_setup.R")
# Load packages
library(tidyverse)
library(cowplot)
library(patchwork)
library(kableExtra)
library(ggrastr)
# Code display options
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=FALSE, echo = TRUE, message=FALSE, warning=FALSE, fig.align="center", fig.retina = 2)
# Load plotting tools
source("../tools/plotting_tools.R")
#Modify the plot theme
theme_set(theme_notebook())
```

------------------------------------------------------------------------

# Fig. 8B - Perfect reads

```{r}
df_read_nearest_down <- read_csv('../../data/seq_data/left_side_TO_libs/df_read_nearest_left.csv')
df_read_nearest_up <- read_csv('../../data/seq_data/right_side_TO_libs/df_read_nearest_right.csv')
df_read_nearest_up_down <- bind_rows(df_read_nearest_up %>% mutate(up_down = 'up'),df_read_nearest_down %>%  mutate(up_down = 'down')) %>% 
  mutate(off_target = ifelse(min_val >500, T, F)) %>% 
  mutate(perfect = ifelse(min_val == 0, T, F))

df_read_nearest_up_down

```

Let's look at both up and down junctions and see how many of the reads (that passed filter) are perfect matches to the targeting oligo deletions we designed.

```{r}

df_read_nearest_up_down_annot <- df_read_nearest_up_down %>% 
  filter(lib >0) %>% 
  mutate(status = paste0('off_',off_target, '_p_',perfect)) %>% 
  mutate(status = fct_relevel(status, 'off_TRUE_p_FALSE','off_FALSE_p_FALSE','off_FALSE_p_TRUE'))
  
plot_perfect_reads <-   ggplot(df_read_nearest_up_down_annot, aes(x = factor(lib), fill = status)) + 
    geom_bar(color = 'black',position = 'fill', linewidth = 0.25) +
    scale_x_discrete(breaks = c('1','2','3'),labels = c('#1 short TFs', '#2 long TFs', '#3 small RNAs'))+
    scale_y_continuous(labels = scales::label_percent())+
    scale_fill_viridis_d(labels = c('Not perfect >500 bp','Not Perfect <500 bp','Perfect match'), option = 'A')  + 
    labs(y = 'Reads', x = 'Library', fill = NULL) + theme(legend.position = 'bottom')

plot_perfect_reads
```

Let's calculate the perfect rate (by read) for each library. Let's separate by the up and downstream junctions to make sure there are no obvious issues.

```{r}
df_read_nearest_up_down_annot %>% group_by(lib, perfect, up_down) %>% summarise(n=n()) %>% pivot_wider(names_from = 'perfect', values_from = 'n', names_prefix = 'perfect_') %>% 
  mutate(frac_perfect = perfect_TRUE / (perfect_TRUE + perfect_FALSE))
```

Looks similar to what the plot tells us.

We can also look at the reads assigned to each target gene and ask what fraction of those reads are "perfect." 

```{r}
df_read_nearest_up_down_annot %>% group_by(lib, gene_id, perfect) %>% summarise(n=n()) %>% pivot_wider(names_from = 'perfect', values_from = 'n', names_prefix = 'perfect_') %>% 
  mutate(frac_perfect = perfect_TRUE / (perfect_TRUE + perfect_FALSE))
  
```

And let's get a summary on that - the mean and median perfect rates (by gene) for each library:

```{r}

df_read_nearest_up_down_annot %>% group_by(lib, gene_id, perfect) %>% summarise(n=n()) %>% pivot_wider(names_from = 'perfect', values_from = 'n', names_prefix = 'perfect_') %>% 
  mutate(frac_perfect = perfect_TRUE / (perfect_TRUE + perfect_FALSE)) %>% group_by(lib) %>% summarise(mean_perf = mean(frac_perfect, na.rm = T), med_perf = median(frac_perfect, na.rm = T))
```

Looks very good. Now let's move on to the target centric analysis.

# Fig. 8C - Targeting oligos

First, let's find all of the targets that have at least 1 perfect read for both the upstream and downstream junctions. We will consider those targets as "found" (score = 2) and targets that do not have perfect reads will be considered "not found." (score = 0)

```{r}

df_TOs <- read_csv("../../../ecoli_ORBIT_paper_final/data/seq_data/right_side_TO_libs/df_TOs.csv")


#Must have at least 1 perfect read
df_read_nearest_all_summary_down <- df_read_nearest_down %>% 
  filter(min_val == 0) %>% 
  group_by(lib, gene_id) %>% 
  summarise( n=n(), min_val = min(min_val, na.rm = T))

df_read_nearest_all_summary_up <- df_read_nearest_up %>% 
  filter(min_val == 0) %>% 
  group_by(lib, gene_id) %>% 
  summarise( n=n(), min_val = min(min_val, na.rm = T))

df_TOs_found_down <- left_join(df_TOs, df_read_nearest_all_summary_down ,by = c('gene'='gene_id', 'lib')) %>% 
  mutate(found_down = ifelse(is.na(n) & is.na(min_val), F, T)) 

df_TOs_found_up_down <- left_join(df_TOs_found_down, df_read_nearest_all_summary_up, suffix = c('_down','_up') ,by = c('gene'='gene_id', 'lib'))%>% 
  mutate(found_up = ifelse(is.na(n_up) & is.na(min_val_up), F, T)) %>% 
  mutate(found_score = found_down + found_up)

df_TOs_found_up_down %>% group_by(lib, found_score) %>% summarise(n=n())

```

Let's express that same data as a plot:

```{r}

plot_tos <- df_TOs_found_up_down %>% 
  filter(lib >0) %>% 
  ggplot(aes(x = factor(lib), fill = factor(found_score))) + geom_bar(position = 'stack', color = 'black', linewidth = 0.25)+
  scale_fill_viridis_d( option = 'A', labels = c('Absent','Present in Up or Dn','Present in both')) + 
  scale_x_discrete(breaks =c('0','1','2','3'),labels = c('#0 oPool','#1 short TFs','#2 long TFs','#3 small RNAs'))+
  labs(fill = 'Mutations found', x = "Mutant library", y = '# Designed targeting oligos') + theme(legend.position = 'bottom')

plot_tos
```

# Fig. 8D - Essential targets.

Now let's look at the absent target genes in more detail. We categorized these targets both programmatically (essential gene lists) and manually (inspecting loci) as essential, perturbing or nonessesnital. See  `upstream_downstream_combined_analysis.Rmd` for details.

```{r}
df_TOs_found_essential_updated <- read_csv("../../data/high_throughput_experiments/df_TOs_found_essential_updated.csv")

df_TOs_found_essential_updated %>% select(lib, gene, found_down, found_up, found_score, essential, essential_score)
```

Let's plot this data simply, for the subset of genes that were totally absent from the ORBIT mutant libraries (found_score = 0).

```{r}

plot_essential <- ggplot(df_TOs_found_essential_updated %>% filter(found_score == 0) %>% mutate(essential = factor(essential, levels = c('essential','probable essential / perturbing','not essential'))), 
       aes(x = essential)) + 
  geom_bar(fill = 'light gray')+
  labs(y = 'absent target genes')

plot_essential
```

# 8E. Example genes from each library

Let's look at a specific target gene example from each of the three libraries: soxS, lysR, and mcaS.

First we need to do some housekeeping to setup the plot - read in gene positions, construct gene arrows, and organize reads by length. 

```{r}
df_genes <- read_delim("../../../ecoli_ORBIT_paper_final/data/seq_data/All_instances_of_Genes_in_Escherichia_coli_K-12_substr._MG1655.txt") %>% 
  mutate(left_pos = `Left-End-Position`, right_pos = `Right-End-Position`) %>% 
  select(-`Left-End-Position`, -`Right-End-Position`) %>% 
  mutate(fwd = ifelse(Direction =='+', T, F))

df_genes_lib <- left_join(df_TOs %>% select(lib, 'gene_id'='gene'), df_genes, by = c('gene_id'='Genes'))


df_poly <- df_genes_lib %>% 
 #filter(right_pos<20000) %>% 
 pivot_longer(c(left_pos, right_pos), values_to = 'x_pos', names_to = 'left_right') %>% 
 mutate( y_pos_bottom = 10,y_pos_top = 100, y_pos_center=31.623) %>% 
 pivot_longer(c(y_pos_bottom,y_pos_top, y_pos_center), values_to = 'y_pos', names_to = 'top_bottom') %>% 
 mutate(x_pos = ifelse(left_right == 'left_pos' & Direction == '-' & top_bottom !='y_pos_center', x_pos + 50, x_pos)) %>% 
 mutate(x_pos = ifelse(left_right == 'right_pos' & Direction == '+' & top_bottom !='y_pos_center', x_pos - 50, x_pos)) %>% 
 mutate(point_num = case_when(left_right == 'left_pos' & top_bottom == 'y_pos_bottom' ~ 1,
left_right == 'left_pos' & top_bottom == 'y_pos_center' ~ 2,
left_right == 'left_pos' & top_bottom == 'y_pos_top' ~ 3,
left_right == 'right_pos' & top_bottom == 'y_pos_top' ~ 4,
left_right == 'right_pos' & top_bottom == 'y_pos_center' ~ 5,
left_right == 'right_pos' & top_bottom == 'y_pos_bottom' ~ 6)) %>% 
 arrange(gene_id, point_num)



 kb_label <- function(x){
 # from s to ns
 lab <- paste(x / 1000)
 }

df_read_nearest_up_down_org <- df_read_nearest_up_down %>% 
  arrange(lib, gene_id, up_down, perfect, width ) %>% 
  group_by(lib, gene_id, up_down, perfect) %>% 
  mutate(read_num = row_number())

df_read_nearest_up_down_org
```

Now let's actually map the reads onto each gene diagram. We will "rasterise" the read data for avoid plotting thousands of individual objects.

```{r}

gene_list <- c('soxS', 'lysR','mcaS')


plot_gene_examples <- ggplot()+
  geom_polygon(data = df_poly %>% filter(lib >0 & gene_id %in% gene_list)%>% mutate(gene_id = factor(gene_id, levels = gene_list)), aes(x = x_pos, y = y_pos, group = gene_id),fill = 'light gray')+
    rasterise(geom_segment(data = df_read_nearest_up_down_org %>% filter(lib >0 & gene_id %in% gene_list) %>% filter(min_val <500) %>% mutate(gene_id = factor(gene_id, levels = gene_list)),
             aes(x = start, xend = end, y = read_num, yend = read_num, color = perfect), alpha = 0.25, linewidth = 0.5), dpi = 2000, dev = "cairo")+
  labs(x = "Genomic position (Kb)", y = 'Read count')+
   scale_x_continuous(labels = kb_label)+
  scale_y_log10(limits =c(1, NA))+
  scale_color_manual(values = c("#440154FF","#21908CFF"))+
  guides(fill = 'none')+
  facet_wrap(~gene_id, scales = 'free', ncol = 1,strip.position = 'top') + guides(color = 'none')

plot_gene_examples
```

And let's get the actual read statistics for this plot:

```{r}
# need to exclude the lysR reads in lib#0

df_read_nearest_up_down %>% filter(gene_id %in% gene_list & lib >0) %>% 
  filter(min_val<=500) %>% 
  group_by(gene_id, perfect) %>% 
  summarise(n=n())%>% 
  pivot_wider(names_from = 'perfect', values_from = 'n', names_prefix = 'perfect_') %>% 
  mutate(n_total = perfect_TRUE + perfect_FALSE, frac_perfect = perfect_TRUE / (perfect_TRUE + perfect_FALSE))

```

# Fig. 8F - long TF deletions overview

For this genome wide overview, we will make three different plots as sort genome "tracks" and then stack them. First are the designed TOs and they are marked for whether mutants were found or not. Next we have all raw reads displayed (relatively transparent) by whether they are perfect or not. Finally, we have a binned histogram of that read data, making it easier to see the fraction of reads that are imperfect across the library.

```{r}
plot_to_track <- ggplot(df_TOs_found_essential_updated %>% filter(lib == 2), aes(x = left_oligo_pos, y= found_up, shape = found_up))+
    geom_point(fill = NA, alpha = 0.5) + 
  scale_shape_manual(values = c(13,21))+
  scale_y_discrete(limits=c(T,F))+
  theme_nothing()

plot_hist <- ggplot(df_read_nearest_up %>% filter(lib==2) %>% mutate(perfect = ifelse(min_val ==0, T, F))) + 
  geom_histogram(aes(x =start, fill =perfect ),binwidth=100000,color = 'black', size = 0.25) +
  #geom_point(data = df_TOs_found_essential_updated %>% filter(found_up ==T)%>% filter(lib==2), aes(x = left_oligo_pos, y = -500), shape =21, fill = NA, alpha =0.5)+
  #geom_point(data = df_TOs_found_essential_updated %>% filter(found_up ==F)%>% filter(lib==2), aes(x = left_oligo_pos, y = -1000 ), shape =13, fill = NA, alpha = 0.5)+
  labs(x = "Genomic position (bp)", y = "Reads (in 100kb bin)", fill = 'Perfect match') +
  scale_fill_manual(values = c("#440154FF","#21908CFF"))+
  scale_x_continuous(labels = Mb_label)
  #ylim(-500,NA)+
  #theme(legend.position = 'bottom')

# plot_jitter <- ggplot(df_read_nearest_up %>% filter(lib==2) %>% mutate(perfect = ifelse(min_val ==0, T, F)), aes(x = start, y = perfect, color = perfect)) + 
#   rasterise(geom_jitter(alpha =0.02, shape = 3, size = 0.5, width = 0), dpi = 5000, dev = "cairo")+
#   scale_color_manual(values = c("#440154FF","#21908CFF"))+
#   #scale_x_continuous(labels = Mb_label)+
#   #scale_y_discrete(labels = c('Off target reads','Perfectly matched reads'))+
#   theme_nothing()

plot_jitter <-ggplot(df_read_nearest_up %>% filter(lib==2) %>% mutate(perfect = ifelse(min_val ==0, T, F)), aes(x = start, y = perfect, color = perfect)) + 
  rasterise(geom_text(label = '|', alpha =0.03, position = position_jitter(width = 0, height = 0.25), size = 2), dpi = 2000, dev = "cairo")+
  scale_color_manual(values = c("#440154FF","#21908CFF"))+
  #scale_x_continuous(labels = Mb_label)+
  #scale_y_discrete(labels = c('Off target reads','Perfectly matched reads'))+
  theme_nothing()

#plot_tos/plot_jitter / plot_hist

#plot_lib_2 <- plot_grid(plot_jitter, plot_hist, align = 'v', axis = 'lr', rel_heights = c(0.3, 1), ncol = 1)

plot_lib_2 <- plot_grid(plot_to_track, plot_jitter, plot_hist, align = 'v', axis = 'lr', rel_heights = c(0.2, 0.33, 1), ncol = 1)


plot_lib_2
```


# Fig. 8G - Upstream and downstream read counts.

Now let's go back and look at the total read counts (upstream + downstream) for all the mutants in the three libraries. We will also use this data for the correlation plots in 8H.

```{r}
df_vars <- read_csv("../../code/supp_figs/df_vars.csv")

df_vars
```

Represent the data as boxplots and show individual genes as points.

```{r}

plot_n_up_down <- ggplot(df_vars %>% filter(lib >0), aes(x = factor(lib), y = n_down + n_up)) + 
  geom_boxplot() + 
  geom_jitter(width = 0.2, height =0, alpha = 0.1, shape =21) + 
  scale_y_log10(limits = c(NA, NA))+
  scale_x_discrete(labels = c('short TFs','long TFs','small RNAs')) +
  labs(x = "Library", y = 'Perfect reads per target\n(Upstream & downstream junctions)')

plot_n_up_down
```

# Fig 8H - Mutant abundance vs. oligo parameters

For all these correlations, only the upstream junction read data was used, since that library had higher counts.

First, let's look at the Twist TO abundance.

```{r}

df_vars %>% 
  filter(lib == 'long TFs') %>% 
  filter(n_up>0 & to_abun>10) %>% 
  mutate(n_up = log(n_up, 10), to_abun = log(to_abun, 10)) %>% 
  lm(n_up ~ to_abun, data =.) %>%
  summary()

plot_to_abun <- ggplot(df_vars %>% filter(to_abun >10 & lib=='long TFs'), aes(x = to_abun, y = n_up)) +
  geom_smooth(method = 'lm')+ 
  geom_point(shape = 21, alpha = 0.25) +
  #scale_x_log10() +
  scale_y_log10()+
  labs(x = 'Reads per oligo', y = 'Perfect reads per target', title = 'Library #2 Long TFs',subtitle = "Reads per target ~ oligo abundance (adj R2 = 0.07, p = 0.00007)")

plot_to_abun
```

Next we examine the homology arm free energy.

```{r}
df_vars %>% 
  filter(lib == 'long TFs') %>% 
  filter(n_up>0) %>% 
  mutate(n_up = log(n_up, 10)) %>% 
  lm(n_up ~ arm_total_duplex, data =.) %>%
  summary()

plot_hom_energy <- ggplot(df_vars %>% filter(lib == 'long TFs') , aes(x = arm_total_duplex, y = n_up)) +
  geom_smooth(method = 'lm')+ 
  geom_point(shape = 21, alpha = 0.25) +
  #scale_x_log10() +
  scale_y_log10()+
  labs(x = 'Homology arm free energy (kcal/mol)', y = 'Perfect reads per target', title = 'Library #2 Long TFs',subtitle = "Reads per target ~ homology arm free energy (adj R2 = 0.12, p = 1.98e-7)")

plot_hom_energy
```

Finally we look at the distance to the origin of replication. We will fit a linear model here just for consistency, but we are only displaying a loess smoothed trend line to show that the ori_dist relationship does not look linear.

```{r}
df_vars %>% 
  filter(lib == 'long TFs') %>% 
  filter(n_up>0) %>% 
  mutate(n_up = log(n_up, 10)) %>% 
  lm(n_up ~ ori_dist, data =.) %>%
  summary()

plot_ori_dist <- ggplot(df_vars %>% filter(lib == 'long TFs') , aes(x = ori_dist, y = n_up)) +
  geom_smooth()+ 
  geom_point(shape = 21, alpha = 0.25) +
  scale_x_continuous(labels = scales::label_comma()) +
  scale_y_log10()+
  labs(x = 'Distance to ori (bp)', y = 'Perfect reads per target', title = 'Library #2 Long TFs',subtitle = "smooth")

plot_ori_dist
```

# Assemble figure

```{r}
theme_set(theme_figure())

plot_top <- plot_grid( plot_perfect_reads, plot_tos,NULL, 
                      NULL, NULL, plot_essential, 
                      align = 'hv', axis = 'lr', rel_widths = c(2,2,1), rel_heights = c(1,0.4), ncol = 3, nrow =2)

plot_top

#save_plot("../../figures/r_pdf_figs/main_figs/fig_8_top.pdf", plot_top, base_width = 5, base_height = 3.25)

plot_lib_2 <- plot_grid(plot_to_track, plot_jitter, plot_hist, align = 'v', axis = 'lr', rel_heights = c(0.2, 0.33, 1), ncol = 1)

plot_middle_pdf <- plot_grid(plot_gene_examples, plot_lib_2, align = 'hv', axis = 'lr', rel_widths = c(1,3), ncol = 2)

plot_middle_pdf

#save_plot("../../figures/r_pdf_figs/main_figs/fig_8_middle.pdf", plot_middle_pdf, base_width = 9, base_height = 3)

plot_bottom <- plot_grid(plot_n_up_down, plot_to_abun, plot_hom_energy, plot_ori_dist, align = 'hv', axis = 'tblr', ncol = 4)

plot_bottom

#save_plot("../../figures/r_pdf_figs/main_figs/fig_8_bottom.pdf", plot_bottom, base_width = 9, base_height = 2)

```

```{r}
sessionInfo()
```
