---
title: "Seq data analysis: Upstream and downstream TO libs"
subtitle: 'E. coli ORBIT 2022'
author: 'Scott H. Saunders'
output: 
  html_document:
    theme: cosmo
    highlight: tango
    code_folding: show
    toc: yes
---

```{r warning=FALSE, message=FALSE}
library(tidyverse)
#library(GenomicAlignments)
library(kableExtra)
library(ggridges)
library(ggbeeswarm)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=FALSE, echo = TRUE, message=FALSE, warning=FALSE, fig.align="center", fig.retina = 2)

source("../tools/plotting_tools.R")
theme_set(theme_notebook())
```

# Notes

This notebook performs the combined junction 1 and junction 2 analysis on the mapped reads for the four ORBIT libraries that were constructed. As a reminder, these different deletion libraries were as follows:

Library #0: galK, hisA, metA, leuD (gold standards) + 23 TF genes (27 targets total)

Library #1: All TF genes < 575 bp (76 targets total)

Library #2: All TF genes > 575 bp (226 targets total)

Library #3: All small RNA genes (90 targets total)

Recall that for each ORBIT mutant a deletion was made and the integrating plasmid was incorporated at that genomic position. Therefore the sequenced junctions bridge the integrating plasmid to the adjacent genome for each mutant.

The junction 1 libraries were processed and analyzed in the notebooks `code/seq_data_processing/left_side_TO_libs_processing.Rmd` and `code/seq_data_processing/left_side_TO_libs_analysis.Rmd`.


The junction 2 libraries were processed and analyzed in the notebooks `code/seq_data_processing/left_side_TO_libs_processing.Rmd` and `code/seq_data_processing/left_side_TO_libs_analysis.Rmd`.

First we will import the junction 1 and 2 reads that were assigned to our target mutants (or off target):

```{r}
df_read_nearest_down <- read_csv('../../data/seq_data/left_side_TO_libs/df_read_nearest_left.csv')
df_read_nearest_up <- read_csv('../../data/seq_data/right_side_TO_libs/df_read_nearest_right.csv')

df_read_nearest_down
df_read_nearest_up

```

We will also go ahead and import our list of designed target mutations (and targeting oligos).

```{r}
df_TOs <- read_csv("../../data/seq_data/right_side_TO_libs/df_TOs.csv")

df_TOs
```


# Read centric analysis

We previously examined the junction 1 and junction 2 libraries individually in their respective analysis notebooks. Here we will combine junction 1 and 2 data to see how many of the reads correspond to designed mutations.

First we will join the datasets and then we can plot the combined read metrics for both libraries:

```{r}
df_read_nearest_up_down <- bind_rows(df_read_nearest_up %>% mutate(up_down = 'up'),df_read_nearest_down %>%  mutate(up_down = 'down')) %>% 
  mutate(off_target = ifelse(min_val >500, T, F)) %>% 
  mutate(perfect = ifelse(min_val == 0, T, F))

df_read_nearest_up_down %>% 
  mutate(status = paste0('off_',off_target, '_p_',perfect)) %>% 
  mutate(status = fct_relevel(status, 'off_TRUE_p_FALSE','off_FALSE_p_FALSE','off_FALSE_p_TRUE')) %>% 
  ggplot(aes(x = lib, fill = status)) + 
    geom_bar(color = 'black',position = 'fill') +
    scale_y_continuous(labels = scales::label_percent())+
    scale_fill_viridis_d(labels = c('Off target / Not Perfect','On Target / Not Perfect','On Target / Perfect'), option = 'A')  + 
    labs(y = 'Reads', x = 'Library', fill = 'Category')
```

We can also examine the read metrics side by side for the junction 1 / 2 libraries (aka down and up).

```{r}
df_read_nearest_up_down %>% 
  mutate(status = paste0('off_',off_target, '_p_',perfect)) %>% 
  mutate(status = fct_relevel(status, 'off_TRUE_p_FALSE','off_FALSE_p_FALSE','off_FALSE_p_TRUE')) %>% 
  ggplot(aes(x = up_down, fill = status)) + 
    geom_bar(color = 'black',position = 'fill') +
  facet_wrap(~lib, ncol = 4)+
    scale_y_continuous(labels = scales::label_percent())+
    scale_fill_viridis_d(labels = c('Off target / Not Perfect','On Target / Not Perfect','On Target / Perfect'), option = 'A')  + 
    labs(y = 'Reads', x = 'Library', fill = 'Category')+
  theme(legend.position = 'bottom')
```

# TO centric analysis

Instead of taking a “read-centric” view, we can instead take a “target-centric” view and ask how many of the TOs we designed actually yielded mutants in our libraries. First we will count up all the reads for each gene_id.

We will find which TOs were "found," meaning that they have at least 1 perfect read in both the junction 1 and junction 2 libraries. 

```{r}

#Must have at least 1 perfect read
df_read_nearest_all_summary_down <- df_read_nearest_down %>% 
  filter(min_val == 0) %>% 
  group_by(lib, gene_id) %>% 
  summarise( n=n(), min_val = min(min_val, na.rm = T))

df_read_nearest_all_summary_up <- df_read_nearest_up %>% 
  filter(min_val == 0) %>% 
  group_by(lib, gene_id) %>% 
  summarise( n=n(), min_val = min(min_val, na.rm = T))

df_TOs_found_down <- left_join(df_TOs, df_read_nearest_all_summary_down ,by = c('gene'='gene_id', 'lib')) %>% 
  mutate(found_down = ifelse(is.na(n) & is.na(min_val), F, T)) 

df_TOs_found_up_down <- left_join(df_TOs_found_down, df_read_nearest_all_summary_up, suffix = c('_down','_up') ,by = c('gene'='gene_id', 'lib'))%>% 
  mutate(found_up = ifelse(is.na(n_up) & is.na(min_val_up), F, T)) %>% 
  mutate(found_score = found_down + found_up)

#write_csv(df_TOs_found_up_down, "df_TOs_found_up_down.csv")

df_TOs_found_up_down %>% select(-left_oligo_pos, -right_oligo_pos, -oligo,-product,-go_term)

```

From this dataset, we can calculate how many of each designed TOs were found represented in the sequenced mutant libraries, with at least a perfect read in both junction 1 and junction 2 (found score = 2).

```{r}

df_TOs_found_up_down %>% 
  group_by(lib, found_score) %>% 
  summarise(n=n()) %>%  
  pivot_wider(names_from = found_score, names_prefix = 'found_', values_from = n,values_fill = 0) %>% 
  mutate(frac_found_2 = found_2 / (found_2 + found_1 + found_0))
```

Overall >89% of TOs were found for each library.

We can also express this in plot form:

```{r}

df_TOs_found_up_down %>% 
  ggplot(aes(x = lib, fill = factor(found_score))) + geom_bar(position = 'fill', color = 'black')+
  scale_fill_viridis_d( option = 'A') + 
  scale_y_continuous(labels = scales::label_percent())

```

The plot in Figure 8 expresses this same data as absolute numbers of TOs in each category (not percentage):

```{r}

df_TOs_found_up_down %>% 
  ggplot(aes(x = factor(lib), fill = factor(found_score))) + geom_bar(position = 'stack', color = 'black')+
  scale_fill_viridis_d( option = 'A', labels = c('Absent','Present in Up or Dn','Present in both')) + 
  scale_x_discrete(breaks =c('0','1','2','3'),labels = c('#0 oPool','#1 short TFs','#2 long TFs','#3 small RNAs'))+
  labs(fill = 'Mutations found', x = "Mutant library", y = '# Designed targeting oligos')
```

# Classifying genes as essential

Although most designed mutations were observed, some were not. One reason why that might be is a mutation cannot be made, because the gene is essential. To test this, we can first download the essential genes lists from Ecocyc. There are three different datasets for LB, the medium used in our experiments.

```{r}
df_lb_1 <- read_delim("../../data/seq_data/318-single-gene-knockouts-exhibit-no-growth-on-LB-Lennox.txt",delim = '\t') %>% 
  mutate(media = 'lennox')

df_lb_2 <- read_delim("../../data/seq_data/358-single-gene-knockouts-exhibit-no-growth-on-Luria-Bertani-broth.txt",delim = '\t') %>% 
  mutate(media = 'luria')

df_lb_3 <- read_delim("../../data/seq_data/614-single-gene-knockouts-exhibit-no-growth-on-LB-enriched.txt",delim = '\t') %>% 
  mutate(media = 'enriched')

df_lb_essential <- bind_rows(df_lb_1, df_lb_2,df_lb_3) %>% 
  mutate(gene=`All-Genes`) %>% 
  select(-`All-Genes`) %>% 
  mutate(essential = T) %>% 
  pivot_wider(names_from = media, names_prefix = "essential_",values_from = essential, values_fill = FALSE) 

df_lb_essential
```

Above, we combined these datasets, and now we can merge this information with our information about which TOs were found.

```{r}
df_TOs_found_essential <- left_join( df_TOs_found_up_down, df_lb_essential, by = 'gene')%>% 
  mutate(essential = ifelse(is.na(essential_lennox),F,T)) %>% 
  mutate(essential_score = essential_lennox + essential_luria + essential_enriched) %>% 
  mutate(essential_score = ifelse(is.na(essential_score), 0, essential_score))

df_TOs_found_essential%>% select(-left_oligo_pos, -right_oligo_pos, -oligo,-product,-go_term, -n_down, -min_val_down, -found_down, -n_up, -min_val_up)

```

We can assume that the more times a gene shows up as essential in the three datasets, the more likely it truly is to be essential. So we can use the 0-3 essential score to assess our confidence in the essentiality. Let's look at this data graphically.

```{r}
ggplot(df_TOs_found_essential %>% mutate(essential_score = factor(essential_score, levels = c('3','2','1','0'))), 
       aes(x =factor(found_score),fill = essential_score)) + 
  geom_bar(color = 'black',position = 'fill') + 
  scale_fill_viridis_d(option = 'A', labels = c('Essential (3x)','Possibly essential (2x)','Possibly essential (1x)','Not essential'))+
  scale_x_discrete(labels = c('Absent','Present in Up or Dn','Present in both'))+
  scale_y_continuous(labels = scales::label_percent())+
  labs(x = 'Mutation category', y = '# Designed targeting oligos')
```

So, this is expected that TOs that were absent (not found), are more likely to be essential genes. Interestingly, some of the TOs not found are not predicted to be essential. Let's examine them manually to see if there is any pattern:


```{r}

df_TOs_found_essential %>% filter(found_score == 0 & essential_score == 0) %>% select(lib, gene, found_score, essential_score,product)

```

There are 13 genes / TOs that fall into this category. Here are some manual observations from looking on Ecocyc:

| gene | lib | Essential or perturbing? | comments                                                                                                                                                                                                                                                                                                   |
|--------------|--------------|--------------|-----------------------------|
| arcA | 2   | no                       | no essential genes nearby. Regulates important metabolic genes, could cause growth defect, but not essential.                                                                                                                                                                                              |
| dcuR | 2   | no                       | no essential genes nearby. Regulates C4 usage, shouldn't be essential in LB.                                                                                                                                                                                                                               |
| fnr  | 2   | no                       | no essential genes nearby. regulates many genes in anaerobic metabolism. Not likely to be essential. Fnr mutant successfully made in IDT library. (n=50)                                                                                                                                                   |
| gntR | 2   | no                       | no essential genes nearby. Regulates gluconate.                                                                                                                                                                                                                                                            |
| metR | 2   | no                       | metR is essential on minimal media because it controls methionine biosynthesis. No LB essential genes nearby.                                                                                                                                                                                              |
| sgrR | 2   | no                       | Involved in sugar transport regulation and adjacent to thiamine transporter. Should not be essential.                                                                                                                                                                                                      |
| uvrY | 2   | no                       | upstream of uvrC then pgsA (which is completely essential 3/3). Seems possible that uvrY ORBIT disrupts regulation of pgsA, but will not classify as perturbing, since it is not the adjacent gene.                                                                                                                                                                                  |
| yheO | 2   | Perturbing               | yheO is part of the tusBCD operon, which makes an important modification to tRNAs. tus mutants have severe phenotypes and have essential score of 1/3. The next gene downstream is rpsL (ribosomal protein), which is totally essential (3/3). Seems possible this mutant disrupts these downstream genes. |
| ffs  | 3   | Essential                | signal recognition particle 4.5s RNA - definitely essential.                                                                                                                                                                                                                                               |
| ohsC | 3   | Essential                | negatively regulates peptide toxin, shoB nearby in genome.                                                                                                                                                                                                                                                 |
| orzP | 3   | no                       | In other E. coli strains it negatively regulates toxic peptide zorP. ZorP doesn't seem to be present in K12.                                                                                                                                                                                               |
| sroA | 3   | no                       | immediately adjacent to sgrR, which also failed in ORBIT and thi genes (thiamine) - overlaps with riboswitch.                                                                                                                                                                                              |
| sroE | 3   | Perturbing               | Directly in between two clearly essential genes. Not clear that this sRNA plays any important role, but seems likely that ORBIT truly disrupted essential locus.                                                                                                                                           |
| tff  | 3   | Perturbing               | Immediately upstream of essential ribosomal subunit. Unclear if sRNA actually plays any role.                                                                                                                                                                                                              |

So, with this information, we can manually update the four genes that are likely perturbing or essential. We can also tweak our classification scheme to be 3 levels - not essential, porbably essential (essential score = 0) / perturbing (manual or essential score = 1-2), or essential (manual or essential score = 3).

```{r}
df_TOs_found_essential_updated <- df_TOs_found_essential %>% 
  mutate(essential = ifelse(essential_score == 3, 'essential',essential)) %>% 
  mutate(essential = ifelse(essential_score %in% c(1,2), 'probable essential / perturbing',essential)) %>% 
  mutate(essential = ifelse(essential_score == 0, 'not essential',essential)) %>% 
  mutate(essential  = ifelse(gene %in% c('yheO','ffs','ohsC','sroE','tff'), 'probable essential / perturbing', essential))

#write_csv(df_TOs_found_essential_updated, "../../data/high_throughput_experiments/df_TOs_found_essential_updated.csv")


df_TOs_found_essential_updated %>% select(lib, gene, found_score, essential, essential_score,product)
```

Finally, we can express this graphically as a percentage or absolute number:

```{r}


ggplot(df_TOs_found_essential_updated %>% mutate(essential = factor(essential, levels = c('essential','probable essential / perturbing','not essential'))), 
       aes(x =factor(found_score),fill = essential)) + 
  geom_bar(color = 'black',position = 'fill') + 
  scale_fill_viridis_d(option = 'A')+ #labels = c('Essential (3x)','Possibly','Not essential')
  scale_x_discrete(labels = c('Absent','Present in Up or Dn','Present in both'))+
  scale_y_continuous(labels = scales::label_percent())+
  labs(x = 'Mutation category', y = '% Designed targeting oligos') + theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggplot(df_TOs_found_essential_updated %>% mutate(essential = factor(essential, levels = c('essential','probable essential / perturbing','not essential'))), 
       aes(x =factor(found_score),fill = essential)) + 
  geom_bar(color = 'black',position = 'stack') + 
  scale_fill_viridis_d(option = 'A')+ #labels = c('Essential (3x)','Possibly','Not essential')
  scale_x_discrete(labels = c('Absent','Present in Up or Dn','Present in both'))+
  labs(x = 'Mutation category', y = '# Designed targeting oligos') + theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

The final plot in Fig. 8 is simplified further to only look at completely absent TOs and just show their essentiality classification:

```{r}
ggplot(df_TOs_found_essential_updated %>% filter(found_score == 0) %>% mutate(essential = factor(essential, levels = c('essential','probable essential / perturbing','not essential'))), 
       aes(x = essential)) + 
  geom_bar(fill = 'light gray')  #labels = c('Essential (3x)','Possibly','Not essential')
  #scale_x_discrete(labels = c('Absent','Present in Up or Dn','Present in both'))+
  #scale_y_continuous(labels = scales::label_percent())+
  #labs(x = 'Mutation category', y = '% Designed targeting oligos') + theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

One follow up question we could ask is, do read counts (relative mutant abundance) follow a trend with the essentiality metric:

```{r}

df_TOs_found_essential_updated %>% pivot_longer(c(n_down, n_up), names_to = 'up_down',values_to = 'up_down_count') %>% 
  mutate(essential = factor(essential, levels = c('essential','probable essential / perturbing','not essential'))) %>% 
  mutate(up_down_count = ifelse(is.na(up_down_count), 0, up_down_count)) %>% 
  ggplot(aes(x = factor(essential), y = up_down_count))+ 
  geom_boxplot(position = position_nudge(x = 0.2),width = 0.1)+
  geom_jitter(shape = 21, width = 0.1, alpha = 0.5,) + 
  facet_wrap(~up_down, ncol = 1)+
  scale_y_log10(oob = scales::squish_infinite, limits = c(0.1, NA))+
  labs(x = 'Mutation category', y = 'Down reads') + theme(axis.text.x = element_text(angle = 45, hjust = 1))


#df_TOs_found_essential_updated %>% filter(essential_score == 3)
```

Indeed, it does seem that essential genes show very few reads, with only a single target showing a couple of reads.

# Next steps

Some of these plot show up in main Figure 8 (`code/main_figs/fig_9=8_to_twist_libs.Rmd`). 

The read count data was also used to model efficiency in several supplemental figures (`code/supp_figs/supp_twist_lib.Rmd`).

```{r}
sessionInfo()
```
